File: ios/TheSharpEdge/Models/Recipe.swift
import Foundation

/// Mirrors api/app/schemas/recipe.py. Unknown keys are ignored on decode;
/// nil optionals are omitted on encode to match the server's exclude_none behaviour.

struct Ingredient: Codable, Hashable, Identifiable {
    var amount: Double        // 0 = to taste → em dash, never scales
    var unit: String
    var name: String
    var note: String?
    var section: String?

    // Stable identity for SwiftUI lists (index-independent within a version).
    var id: String { "\(section ?? "")|\(name)|\(unit)|\(amount)" }

    init(amount: Double = 0, unit: String = "", name: String, note: String? = nil, section: String? = nil) {
        self.amount = amount
        self.unit = unit
        self.name = name
        self.note = note
        self.section = section
    }

    enum CodingKeys: String, CodingKey { case amount, unit, name, note, section }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        amount = try c.decodeIfPresent(Double.self, forKey: .amount) ?? 0
        unit = try c.decodeIfPresent(String.self, forKey: .unit) ?? ""
        name = try c.decode(String.self, forKey: .name)
        note = try c.decodeIfPresent(String.self, forKey: .note)
        section = try c.decodeIfPresent(String.self, forKey: .section)
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(amount, forKey: .amount)
        try c.encode(unit, forKey: .unit)
        try c.encode(name, forKey: .name)
        try c.encodeIfPresent(note, forKey: .note)
        try c.encodeIfPresent(section, forKey: .section)
    }
}

struct Step: Codable, Hashable, Identifiable {
    var text: String
    var timerSeconds: Int?

    var id: String { text }

    init(text: String, timerSeconds: Int? = nil) {
        self.text = text
        self.timerSeconds = timerSeconds
    }

    enum CodingKeys: String, CodingKey {
        case text
        case timerSeconds = "timer_seconds"
    }
}

struct RecipeCard: Codable, Hashable, Identifiable {
    var slug: String
    var title: String
    var category: String
    var meta: String?
    var baseYield: Int
    var yieldWord: String
    var gf: Bool
    var noscale: Bool
    var status: String

    var id: String { slug }

    enum CodingKeys: String, CodingKey {
        case slug, title, category, meta
        case baseYield = "base_yield"
        case yieldWord = "yield_word"
        case gf, noscale, status
    }
}

struct VersionOut: Codable, Hashable, Identifiable {
    var id: UUID
    var version: Int
    var label: String?
    var ingredients: [Ingredient]
    var steps: [Step]
    var notes: [String]
    var isCurrent: Bool
    var createdAt: Date

    enum CodingKeys: String, CodingKey {
        case id, version, label, ingredients, steps, notes
        case isCurrent = "is_current"
        case createdAt = "created_at"
    }
}

struct VersionSummary: Codable, Hashable, Identifiable {
    var id: UUID
    var version: Int
    var label: String?
    var isCurrent: Bool
    var createdAt: Date

    enum CodingKeys: String, CodingKey {
        case id, version, label
        case isCurrent = "is_current"
        case createdAt = "created_at"
    }
}

struct RecipeFull: Codable, Hashable, Identifiable {
    var slug: String
    var title: String
    var category: String
    var meta: String?
    var baseYield: Int
    var yieldWord: String
    var gf: Bool
    var noscale: Bool
    var status: String
    var source: String?
    var currentVersion: VersionOut

    var id: String { slug }

    enum CodingKeys: String, CodingKey {
        case slug, title, category, meta
        case baseYield = "base_yield"
        case yieldWord = "yield_word"
        case gf, noscale, status, source
        case currentVersion = "current_version"
    }
}

/// POST body — creates a recipe and its version 1.
/// Differs from RecipeUpdate in exactly one way that matters: the slug is
/// caller-supplied and permanent. QR codes are printed against it (CLAUDE.md §5),
/// so it is generated once via /parse/slug and confirmed before save.
struct RecipeCreate: Codable, Identifiable {
    /// Presentation identity only — `sheet(item:)` needs it. Never sent on the wire.
    var id: String { slug.isEmpty ? "new" : slug }

    var slug: String
    var title: String
    var category: String
    var meta: String?
    var baseYield: Int
    var yieldWord: String
    var gf: Bool
    var noscale: Bool
    var source: String?
    var status: String
    var label: String?
    var ingredients: [Ingredient]
    var steps: [Step]
    var notes: [String]

    init(slug: String = "", title: String = "", category: String = Category.order[0],
         meta: String? = nil, baseYield: Int = 4, yieldWord: String = "servings",
         gf: Bool = false, noscale: Bool = false, source: String? = nil,
         status: String = "draft", label: String? = nil,
         ingredients: [Ingredient] = [], steps: [Step] = [], notes: [String] = []) {
        self.slug = slug
        self.title = title
        self.category = category
        self.meta = meta
        self.baseYield = baseYield
        self.yieldWord = yieldWord
        self.gf = gf
        self.noscale = noscale
        self.source = source
        self.status = status
        self.label = label
        self.ingredients = ingredients
        self.steps = steps
        self.notes = notes
    }

    enum CodingKeys: String, CodingKey {
        case slug, title, category, meta
        case baseYield = "base_yield"
        case yieldWord = "yield_word"
        case gf, noscale, source, status, label
        case ingredients, steps, notes
    }
}

/// PUT body — appends a new version. Slug never changes.
struct RecipeUpdate: Codable {
    var title: String?
    var category: String?
    var meta: String?
    var baseYield: Int?
    var yieldWord: String?
    var gf: Bool?
    var noscale: Bool?
    var source: String?
    var status: String?
    var label: String?
    var ingredients: [Ingredient]
    var steps: [Step]
    var notes: [String]

    enum CodingKeys: String, CodingKey {
        case title, category, meta
        case baseYield = "base_yield"
        case yieldWord = "yield_word"
        case gf, noscale, source, status, label
        case ingredients, steps, notes
    }
}

File: api/app/schemas/recipe.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

ALLOWED_UNITS = {"g", "ml", "cup", "tbsp", "tsp", "lb", "oz", ""}


class Ingredient(BaseModel):
    model_config = ConfigDict(extra="allow")  # section, note ride along

    amount: float = 0  # 0 = to taste → em dash, never scales
    unit: str = ""
    name: str
    note: str | None = None
    section: str | None = None


class Step(BaseModel):
    model_config = ConfigDict(extra="allow")

    text: str
    timer_seconds: int | None = None


class PageRef(BaseModel):
    """Notebook page mapping — feeds the cards.pdf glue-in index (§10)."""

    model_config = ConfigDict(from_attributes=True)

    page_number: int = Field(ge=1)
    section: str | None = None


class RecipeCard(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    slug: str
    title: str
    category: str
    meta: str | None = None
    base_yield: int
    yield_word: str
    gf: bool
    noscale: bool
    status: str
    tags: list[str] = []


class VersionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    version: int
    label: str | None = None
    ingredients: list[Ingredient]
    steps: list[Step]
    notes: list[str]
    is_current: bool
    created_at: datetime


class RecipeFull(RecipeCard):
    source: str | None = None
    pages: list[PageRef] = []
    current_version: VersionOut


class VersionSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    version: int
    label: str | None = None
    is_current: bool
    created_at: datetime


class RecipeCreate(BaseModel):
    slug: str = Field(pattern=r"^[a-z0-9][a-z0-9-]*$")
    title: str
    category: str
    meta: str | None = None
    base_yield: int = Field(ge=1)
    yield_word: str = "servings"
    gf: bool = False
    noscale: bool = False
    source: str | None = None
    status: str = "active"
    label: str | None = None
    ingredients: list[Ingredient] = []
    steps: list[Step] = []
    notes: list[str] = []
    tags: list[str] = []
    pages: list[PageRef] = []


class RecipeUpdate(BaseModel):
    """PUT — appends a new version; recipe fields update in place. Slug never changes."""

    title: str | None = None
    category: str | None = None
    meta: str | None = None
    base_yield: int | None = Field(default=None, ge=1)
    yield_word: str | None = None
    gf: bool | None = None
    noscale: bool | None = None
    source: str | None = None
    status: str | None = None
    label: str | None = None
    ingredients: list[Ingredient]
    steps: list[Step]
    notes: list[str] = []
    tags: list[str] | None = None  # None = leave unchanged
    pages: list[PageRef] | None = None  # None = leave unchanged


class CookSessionCreate(BaseModel):
    started_at: datetime | None = None  # None = finished_at (untimed log)
    scaled_yield: int = Field(ge=1)
    notes: str | None = Field(default=None, max_length=500)


class CookSessionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    started_at: datetime
    finished_at: datetime
    scaled_yield: int
    notes: str | None = None


class ScaleRequest(BaseModel):
    target_yield: int = Field(ge=1)


class ScaledIngredient(Ingredient):
    scaled_amount: float
    display: str


class ScaleResponse(BaseModel):
    slug: str
    base_yield: int
    target_yield: int
    yield_word: str
    ingredients: list[ScaledIngredient]
